# XML, HTML, and Web Scraping

JSON and XML are two different ways to represent hierarchical data. Which one is better? There are lots of articles online which discuss similarities and differences between JSON and XML and their advantages and disadvantages. Both formats are still in current usage, so it is good to be familiar with both. However, JSON is more common, so we'll focus on working with JSON representations of hierarchical data.

The reading covered an example of using Beautiful Soup to parse XML. Rather than doing another example XML now, we'll skip straight to scraping HTML from a webpage. Both HTML and XML can be parsed in a similar way with Beautiful Soup.

In [ ]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

In [ ]:
#pip install beautifulsoup4

## Scraping an HTML table with Beautiful Soup

Open the URL https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population and scroll down until you see a table of the cities in the U.S. with population over 100,000 (as of Jul 1, 2022). We'll use Beautiful Soup to scrape information from this table.

Read in the HTML from the ULR using the `requests` library.

In [ ]:
# YOUR CODE HERE

URL = "https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population"
HEADERS = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"}

response = requests.get(URL, headers=HEADERS)

Use Beautiful Soup to parse this string into a tree called `soup`

In [ ]:
# YOUR CODE HERE

#response = requests.get("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")
from bs4 import BeautifulSoup
soup = BeautifulSoup(response.text, "html.parser")


To find an HTML tag corresponding to a specific element on a webpage, right-click on it and choose "Inspect element". Go to the cities table Wikipedia page and do this now.

You should find that the cities table on the Wikipedia page corresponds to the element

```
<table class="wikitable sortable jquery-tablesorter" style="text-align:center">
```

There are many `<table>` tags on the page.

In [ ]:
len(soup.find_all("table"))

10

We can use attributes like `class=` and `style=` to narrow down the list.

In [ ]:
len(soup.find_all("table",
                  attrs={
                      "class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center",
                      "style": "text-align:right"}
                  ))

1

At this point, you can manually inspect the tables on the webpage to find that the one we want is the first one (see `[0]` below). We'll store this as `table`.

In [ ]:
table = soup.find_all("table",
                  attrs={
                       "class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center",
                       "style": "text-align:right"}
                  )[0]

**Now you will write code to scrape the information in `table` to create a Pandas data frame with one row for each city and columns for: city, state, population (2022 estimate), and 2020 land area (sq mi).** Refer to the Notes/suggestions below as you write your code. A few Hints are provided further down, but try coding first before looking at the hints.

Notes/suggestions:

- Use as a guide the code from the reading that produced the data frame of Statistics faculty
- Inspect the page source as you write your code
- You will need to write a loop to get the information for all cities, but you might want to try just scraping the info for New York first
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `.get_text(strip = True)` instead of `.text`
- Don't forget to convert to a Pandas Data Frame; it should have 333 rows and 4 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it --- e.g., what is the population density for all cities in CA? --- then you would need to clean the data first (to clean strings and convert to quantitative). (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)

In [ ]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [ ]:
city = table.find_all("tr")
len(city)

348

In [ ]:
city = table.find_all("tr")[2]
city

<tr>
<td style="background-color:#cfecec"><i><a href="/wiki/New_York_City" title="New York City">New York</a></i><sup class="reference" id="cite_ref-5"><a href="#cite_note-5"><span class="cite-bracket">[</span>c<span class="cite-bracket">]</span></a></sup>
</td>
<td><a href="/wiki/New_York_(state)" title="New York (state)">NY</a>
</td>
<td style="text-align:right;">8,478,072
</td>
<td style="text-align:right;">8,804,190
</td>
<td style="text-align:right;"><span data-sort-value="2999630000000000000♠" style="display:none"></span><span style="color:red">−3.70%</span>
</td>
<td style="text-align:right;">300.5
</td>
<td style="text-align:right;">778.3
</td>
<td style="text-align:right;">29,298
</td>
<td style="text-align:right;">11,312
</td>
<td><small><span class="geo-inline"><style data-mw-deduplicate="TemplateStyles:r1156832818">.mw-parser-output .geo-default,.mw-parser-output .geo-dms,.mw-parser-output .geo-dec{display:inline}.mw-parser-output .geo-nondefault,.mw-parser-output .geo-mult

In [ ]:
cells = city.find_all("td")
cells

[<td style="background-color:#cfecec"><i><a href="/wiki/New_York_City" title="New York City">New York</a></i><sup class="reference" id="cite_ref-5"><a href="#cite_note-5"><span class="cite-bracket">[</span>c<span class="cite-bracket">]</span></a></sup>
 </td>,
 <td><a href="/wiki/New_York_(state)" title="New York (state)">NY</a>
 </td>,
 <td style="text-align:right;">8,478,072
 </td>,
 <td style="text-align:right;">8,804,190
 </td>,
 <td style="text-align:right;"><span data-sort-value="2999630000000000000♠" style="display:none"></span><span style="color:red">−3.70%</span>
 </td>,
 <td style="text-align:right;">300.5
 </td>,
 <td style="text-align:right;">778.3
 </td>,
 <td style="text-align:right;">29,298
 </td>,
 <td style="text-align:right;">11,312
 </td>,
 <td><small><span class="geo-inline"><style data-mw-deduplicate="TemplateStyles:r1156832818">.mw-parser-output .geo-default,.mw-parser-output .geo-dms,.mw-parser-output .geo-dec{display:inline}.mw-parser-output .geo-nondefault,.mw-

In [ ]:
#city
name = cells[0].get_text(strip= True)
name

'New York[c]'

In [ ]:
#state
state = cells[1].get_text(strip= True)
state

'NY'

In [ ]:
 #city, state, population (2022 estimate), and 2020 land area (sq mi)

In [ ]:
#population
population = cells[2].get_text(strip= True)
population

'8,478,072'

In [ ]:
#land area
area = cells[5].get_text(strip= True)
area

'300.5'

In [ ]:
# initialize an empty list
rows = []

# iterate over all rows in the city table
for city in table.find_all("tr")[2:]:

    # Get all the cells (<td>) in the row.
    cells = table.find_all("td")

#see code above
    name = cells[0].find("a")
    name = name.text

    state = cells[1].get_text(strip= True)
    state

    population = cells[2].get_text(strip= True)
    population

    area = cells[5].get_text(strip= True)
    area

    # Append this data.
    rows.append({
        "city": name,
        "state": state,
        "population": population,
        "area": area
    })

In [ ]:
pd.DataFrame(rows)

city state population   area
0    New York    NY  8,478,072  300.5
1    New York    NY  8,478,072  300.5
2    New York    NY  8,478,072  300.5
3    New York    NY  8,478,072  300.5
4    New York    NY  8,478,072  300.5
..        ...   ...        ...    ...
341  New York    NY  8,478,072  300.5
342  New York    NY  8,478,072  300.5
343  New York    NY  8,478,072  300.5
344  New York    NY  8,478,072  300.5
345  New York    NY  8,478,072  300.5

[346 rows x 4 columns]

Hints:

- Each city is a row in the table; find all the `<tr>` tags to find all the cities
- Look for the `<td>` tag to see table entries within a row
- The rank column is represented by `<th>` tags, rather than `<td>` tags. So within a row, the first (that is, `[0]`) `<td>` tag corresponds to the city name.

## Aside: Scraping an HTML table with Pandas



The Pandas command `read_html` can be used to scrape information from an HTML table on a webpage.

We can call `read_html` on the URL.

In [ ]:
URL = "https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population"
HEADERS = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"}

response = requests.get(URL, headers=HEADERS)

In [ ]:
response.text

'<!DOCTYPE html>\n<html class="client-nojs vector-feature-language-in-header-enabled vector-feature-language-in-main-page-header-disabled vector-feature-page-tools-pinned-disabled vector-feature-toc-pinned-clientpref-1 vector-feature-main-menu-pinned-disabled vector-feature-limited-width-clientpref-1 vector-feature-limited-width-content-enabled vector-feature-custom-font-size-clientpref-1 vector-feature-appearance-pinned-clientpref-1 vector-feature-night-mode-enabled skin-theme-clientpref-day vector-sticky-header-enabled vector-toc-available" lang="en" dir="ltr">\n<head>\n<meta charset="UTF-8">\n<title>List of United States cities by population - Wikipedia</title>\n<script>(function(){var className="client-js vector-feature-language-in-header-enabled vector-feature-language-in-main-page-header-disabled vector-feature-page-tools-pinned-disabled vector-feature-toc-pinned-clientpref-1 vector-feature-main-menu-pinned-disabled vector-feature-limited-width-clientpref-1 vector-feature-limited

In [ ]:
from io import StringIO

df = pd.read_html(StringIO(response.text))

df

[                    Population tables of U.S. cities
 0  The skyline of New York City, the most populou...
 1                                             Cities
 2  PopulationAreaDensityEthnic identityForeign-bo...
 3                                        Urban areas
 4             Populous cities and metropolitan areas
 5                                 Metropolitan areas
 6  184 combined statistical areas935 core-based s...
 7                                        Megaregions
 8  Related population listsNorth American metro a...
 9                                                vte,
     0                               1
 0 NaN                   State capital
 1 NaN            State's largest city
 2 NaN  State capital and largest city
 3 NaN                 Federal capital,
             City  ST 2024 estimate 2020 census  Change 2020 land area          \
             City  ST 2024 estimate 2020 census  Change            mi2     km2   
 0    New York[c]  NY       8478072     88041

However, this scrapes all the tables on the webpage, not just the one we want. As with Beautiful Soup, we can narrow the search by specifying the table attributes.

In [ ]:
df2 = pd.read_html(StringIO(response.text), attrs = {"class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center",
                       "style": "text-align:right"})
df2

[            City  ST 2024 estimate 2020 census  Change 2020 land area          \
             City  ST 2024 estimate 2020 census  Change            mi2     km2   
 0    New York[c]  NY       8478072     8804190  −3.70%          300.5   778.3   
 1    Los Angeles  CA       3878704     3898747  −0.51%          469.5  1216.0   
 2        Chicago  IL       2721308     2746388  −0.91%          227.7   589.7   
 3        Houston  TX       2390125     2304580  +3.71%          640.4  1658.6   
 4        Phoenix  AZ       1673164     1608139  +4.04%          518.0  1341.6   
 ..           ...  ..           ...         ...     ...            ...     ...   
 341      Deltona  FL        100513       93692  +7.28%           37.3    96.6   
 342  Federal Way  WA        100252      101030  −0.77%           22.3    57.8   
 343   San Angelo  TX        100159       99893  +0.27%           59.7   154.6   
 344        Tracy  CA        100136       93000  +7.67%           25.9    67.1   
 345      Sunris

This still returns 3 tables. As we remarked above, the table that we want is the first one (see `[0]` below).

In [ ]:
#df_cities2 = pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {'class': 'wikitable sortable', "style": "text-align:center"})[0]
#df_cities2

df_cities2 = pd.read_html(StringIO(response.text), attrs = {"class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center",
                       "style": "text-align:right"})[0]
df_cities2

City  ST 2024 estimate 2020 census  Change 2020 land area          \
            City  ST 2024 estimate 2020 census  Change            mi2     km2   
0    New York[c]  NY       8478072     8804190  −3.70%          300.5   778.3   
1    Los Angeles  CA       3878704     3898747  −0.51%          469.5  1216.0   
2        Chicago  IL       2721308     2746388  −0.91%          227.7   589.7   
3        Houston  TX       2390125     2304580  +3.71%          640.4  1658.6   
4        Phoenix  AZ       1673164     1608139  +4.04%          518.0  1341.6   
..           ...  ..           ...         ...     ...            ...     ...   
341      Deltona  FL        100513       93692  +7.28%           37.3    96.6   
342  Federal Way  WA        100252      101030  −0.77%           22.3    57.8   
343   San Angelo  TX        100159       99893  +0.27%           59.7   154.6   
344        Tracy  CA        100136       93000  +7.67%           25.9    67.1   
345      Sunrise  FL        100128       97335  +2.87%           16.2    42.0   

    2020 density                                      Location  
           / mi2  / km2                               Location  
0          29298  11312    40°40′N 73°56′W / 40.66°N 73.94°W  
1           8304   3206  34°01′N 118°25′W / 34.02°N 118.41°W  
2          12061   4657    41°50′N 87°41′W / 41.84°N 87.68°W  
3           3599   1390    29°47′N 95°23′W / 29.79°N 95.39°W  
4           3105   1199  33°34′N 112°05′W / 33.57°N 112.09°W  
..           ...    ...                                    ...  
341         2512    970    28°55′N 81°13′W / 28.91°N 81.21°W  
342         4530   1750  47°19′N 122°20′W / 47.31°N 122.34°W  
343         1673    646  31°26′N 100°27′W / 31.44°N 100.45°W  
344         3591   1386  37°44′N 121°27′W / 37.73°N 121.45°W  
345         6008   2320    26°10′N 80°16′W / 26.17°N 80.26°W  

[346 rows x 10 columns]

Wait, that seemed much easier than using Beautiful Soup, and it returned a data frame, and we even got for free some formatting like removing the commas from the population! Why didn't we just use `read_html` in the first place? It's true the `read_html` works well when scraping information from an HTML *table*. Unfortunately, you often want to scrape information from a webpage that isn't conveniently stored in an HTML table, in which case `read_html` won't work. (It only searches for `<table>`, `<th>`, `<tr>`, and `<td>` tags, but there are many other HTML tags.) Though Beautiful Soup is not as simple as `read_html`, it is more flexible and thus more widely applicable.

## Scraping information that is NOT in a `<table>` with Beautiful Soup

The Cal Poly course catalog http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory contains a list of courses offered by the Statistics department. **You will scrape this website to obtain a Pandas data frame with one row for each DATA or STAT course and two columns: course name and number (e.g, DATA 301. Introduction to Data Science) and term typically offered (e.g., Term Typically Offered: F, W, SP).**

Note: Pandas `read_html` is not help here since the courses are not stored in a `<table>.`

In [ ]:
pd.read_html("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")

[                                        Program name   Program type
 0                              Actuarial Preparation          Minor
 1  Cross Disciplinary Studies Minor in Bioinforma...          Minor
 2   Cross Disciplinary Studies Minor in Data Science          Minor
 3                                         Statistics  BS, MS, Minor]


Notes/suggestions:


- Inspect the page source as you write your code
- The courses are not stored in a `<table>`. How are they stored?
- You will need to write a loop to get the information for all courses, but you might want to try just scraping the info for DATA 100 first
- What kind of tag is the course name stored in? What is the `class` of the tag?
- What kind of tag is the quarter(s) the course is offered stored in? What is the `class` of the tag? Is this the only tag of this type with the class? How will you get the one you want?
- You don't have to remove the number of units (e.g., 4 units) from the course name and number, but you can try it if you want
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `get_text(strip = True)` instead of `text`
- Don't forget to convert to a Pandas Data Frame; it should have 74 rows and 2 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it then you might need to clean the data first. (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)



In [ ]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [ ]:
# Read in the HTML from the URL
URL = "http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory"
HEADERS = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"}
response = requests.get(URL, headers=HEADERS)

In [ ]:
# Parse the HTML with Beautiful Soup
soup = BeautifulSoup(response.text, "html.parser")

In [ ]:
# First, find all course blocks
courses = soup.find_all("div", {"class": "courseblock"})
#courses #ran this before, gave the correct output but was super long so put it in hash tags for now

In [ ]:
# Looking at the first course block only
course = courses[0]
course

<div class="courseblock">
<p class="courseblocktitle"><strong>DATA 100. Data Science for All I.
<span class="courseblockhours">4 units
</span></strong></p><div class="noindent courseextendedwrap">
<p class="noindent">Term Typically Offered: F, W, SP</p><p class="noindent">2020-21 or later catalog: GE Area B4</p><p class="noindent">2019-20 or earlier catalog: GE Area B4</p><p>Prerequisite: <a class="bubblelink code" href="/search/?P=MATH%20115" onclick="return showCourse(this, 'MATH 115');" title="MATH 115">MATH 115</a>, <a class="bubblelink code" href="/search/?P=MATH%20116" onclick="return showCourse(this, 'MATH 116');" title="MATH 116">MATH 116</a>, <a class="bubblelink code" href="/search/?P=MATH%20118" onclick="return showCourse(this, 'MATH 118');" title="MATH 118">MATH 118</a>, or Appropriate Math Placement Level.</p></div>
<div class="courseblockdesc">
<p>Basic approaches for answering questions using data.  Emphasis on working with tabular data in spreadsheet software to provide

In [ ]:
# Find the course title
course_title = course.find("p", {"class": "courseblocktitle"})
course_title

<p class="courseblocktitle"><strong>DATA 100. Data Science for All I.
<span class="courseblockhours">4 units
</span></strong></p>

In [ ]:
#I want just the title
course_title.get_text(strip=True)

'DATA\xa0100. Data Science for All I.4 units'

In [ ]:
# Find the units
units = course_title.find("span", {"class": "courseblockhours"})
units #didn't end up needing this code, looked at hints to remove units

<span class="courseblockhours">4 units
</span>

In [ ]:
# Find the div with class="noindent courseextendedwrap" for F,W,S offering
extended_wrap = course.find("div", {"class": "noindent courseextendedwrap"})
extended_wrap

<div class="noindent courseextendedwrap">
<p class="noindent">Term Typically Offered: F, W, SP</p><p class="noindent">2020-21 or later catalog: GE Area B4</p><p class="noindent">2019-20 or earlier catalog: GE Area B4</p><p>Prerequisite: <a class="bubblelink code" href="/search/?P=MATH%20115" onclick="return showCourse(this, 'MATH 115');" title="MATH 115">MATH 115</a>, <a class="bubblelink code" href="/search/?P=MATH%20116" onclick="return showCourse(this, 'MATH 116');" title="MATH 116">MATH 116</a>, <a class="bubblelink code" href="/search/?P=MATH%20118" onclick="return showCourse(this, 'MATH 118');" title="MATH 118">MATH 118</a>, or Appropriate Math Placement Level.</p></div>

In [ ]:
# Find all tags with class="noindent" inside this div
# The first one is "Term Typically Offered"
term = extended_wrap.find("p", {"class": "noindent"})
term_offered = term.get_text(strip=True)
term_offered

'Term Typically Offered: F, W, SP'

In [ ]:
# Now putting it all together

rows = []

# Iterate over all courses
for course in soup.find_all("div", {"class": "courseblock"}):

    # Find the course name
    course_title = course.find("p", {"class": "courseblocktitle"})

     # Remove the span with units, used the hints below to get this
    span = course_title.find("span", {"class": "courseblockhours"})
    if span:
        span.extract()

    course_name = course_title.get_text(strip=True)

    #F,W,S offering
    extended_wrap = course.find("div", {"class": "noindent courseextendedwrap"})

    #need the very first one to get term typically offered
    term = extended_wrap.find("p", {"class": "noindent"})
    term_offered = term.get_text(strip=True)

    # Append data
    rows.append({
        "course": course_name,
        "term_offered": term_offered
    })

#DataFrame
df_courses = pd.DataFrame(rows)
df_courses

course  \
0                   DATA 100. Data Science for All I.   
1             DATA 301. Introduction to Data Science.   
2          DATA 401. Data Science Process and Ethics.   
3   DATA 402. Mathematical Foundations of Data Sci...   
4         DATA 403. Data Science Projects Laboratory.   
..                                                ...   
69             STAT 551. Statistical Learning with R.   
70           STAT 566. Graduate Consulting Practicum.   
71                STAT 570. Selected Advanced Topics.   
72          STAT 590. Graduate Seminar in Statistics.   
73                                  STAT 599. Thesis.   

                        term_offered  
0   Term Typically Offered: F, W, SP  
1   Term Typically Offered: F, W, SP  
2          Term Typically Offered: F  
3          Term Typically Offered: F  
4          Term Typically Offered: F  
..                               ...  
69         Term Typically Offered: F  
70        Term Typically Offered: SP  
71       Term Typically Offered: TBD  
72  Term Typically Offered: F, W, SP  
73       Term Typically Offered: TBD  

[74 rows x 2 columns]

Hints:

- Each course is represented by a `<div>` with `class=courseblock`, so you can find all the courses with `soup.find_all("div", {"class": "courseblock"})`
- The course name is in a `<p>` tag with `class=courseblocktitle`, inside a `<strong>` tag. (Though I don't think we need to find the strong tag here.)
- The term typically offered is in `<p>` tag with `class=noindent`. However, there are several tags with this class; term typically offered is the first one.
- If you want to use Beautiful Soup to remove the course units (e.g., 4 units), find the `<span>` tag within the course name tag and `.extract()` this span tag